updates: Redesign available and up-to-date pages and autoupdates

skobyda opened the pull request:
Fixes: #14749
Fixes: #14723

Use PF4 cards for updates page. Redesign and refactor autoupdates.

### Autoupdates
![Screenshot from 2020-11-20 11-04-30](https://user-images.githubusercontent.com/42733240/99790264-2422a780-2b24-11eb-9b67-59217867dd5a.png)
![Screenshot from 2020-11-20 11-04-44](https://user-images.githubusercontent.com/42733240/99790273-271d9800-2b24-11eb-9ee1-299d5a45d904.png)


### Available
![Screenshot from 2020-11-20 11-03-47](https://user-images.githubusercontent.com/42733240/99790005-c2fad400-2b23-11eb-9e0d-59c875ad44d7.png)
.
.
.
![Screenshot from 2020-11-20 11-04-01](https://user-images.githubusercontent.com/42733240/99790017-c8581e80-2b23-11eb-9c33-84f566afff5d.png)


### Up to date
![Screenshot from 2020-11-20 11-37-19](https://user-images.githubusercontent.com/42733240/99790708-c93d8000-2b24-11eb-9f46-3370d8e13d08.png)

marusak commented on pkg/packagekit/autoupdates.jsx:
```diff
@@ -249,67 +248,128 @@ function getBackend(forceReinit) {
-export default class AutoUpdates extends React.Component {
+export const AutoUpdatesBody = (props) => {
+    const { enabled, type, day, time } = props;
+    const days = {
+        "": "everyday",
+        mon: "Monday",
+        tue: "Tuesday",
+        wed: "Wednesday",
+        thu: "Thursday",
+        fri: "Friday",
+        sat: "Saturday",
+        sun: "Sunday"
+    };
+
+    let str;
+    if (enabled) {
+        str = type == "security" ? _("Security updates ") : _("Updates ");
+        str += day == "" ? cockpit.format(_("will be applied $0 at $1"), days[day], time) : cockpit.format(_("will be applied every $0 at $1"), days[day], time);
```
- `days` are not localized.
- space at the end of `Security updates ` and `Updates ` are super likely to get broken by translators
- But more importantly translating `will be applied $0 at $1` is difficult to translate without any context. `every Monday` in slovak is `každý pondelok` while `every Wednesday` is `každú stredu`. Also translating just `Wednesday` (as in the list above) I wouls translate `streda` and not `stredu` so the resulting string would be `budú aplikované každý streda` which sounds like you speak Slovak for the first time.

Either moment offers something for this or we need to do this in better way. One option is to list all the options for all 8 days and twice for `every` or without. 16 strings, not ideal... Other option would be to rethink the format of that string and write it in such way that it would work with neutral forms.

skobyda replied:
I cannot think of a format of that string in a neutral form. Perhaps we could use that object and add every in front of it? Like:     
```
const days = {
        "": _("everyday"),
        mon: _("every Monday"),
        tue: _("every Tuesday"),
        wed: _("every Wednesday"),
        thu: _("every Thursday"),
        fri: _("every Friday"),
        sat: _("every Saturday"),
        sun: _("every Sunday"),
    };
```

marusak reviewed (changes requested):
Just the first thing I saw :)

andreasn commented on pkg/packagekit/autoupdates.jsx:
```diff
@@ -320,41 +380,53 @@ export default class AutoUpdates extends React.Component {
+                    }}>
+                {_("Edit")}
+            </Button>
+            <Modal position="top" variant="medium" id="automatic-updates-dialog" isOpen={this.state.showModal} onClose={() => this.setState({ showModal: false })}
```
Dialog feels extremely wide. Would the small variant work better maybe?
https://www.patternfly.org/v4/components/modal#small

andreasn reviewed (changes requested):
The two cards don't seem to agree on what's a good uniform hight.
![image](https://user-images.githubusercontent.com/209895/99795572-45879180-2b2c-11eb-829b-040ed0248c93.png)

Smaller sizes has some overflow issues.
![image](https://user-images.githubusercontent.com/209895/99795656-69e36e00-2b2c-11eb-8469-c33846e722ab.png)

KKoukiou commented on pkg/packagekit/autoupdates.jsx:
```diff
@@ -249,67 +248,128 @@ function getBackend(forceReinit) {
+    const { enabled, type, day, time } = props;
+    const days = {
+        "": "everyday",
+        mon: "Monday",
```
Needs translation

KKoukiou commented on pkg/packagekit/autoupdates.jsx:
```diff
@@ -249,67 +248,128 @@ function getBackend(forceReinit) {
+                           enabled={enabled}
+                           label={_("Security updates only")}
+                           id="security-updates"
+                           name="radio-security-updates" />
```
The name field of the radio buttons that belong to one group should be the same. See https://www.w3schools.com/tags/att_input_type_radio.asp

skobyda replied:
Fixed

KKoukiou commented on pkg/packagekit/autoupdates.jsx:
```diff
@@ -320,41 +380,53 @@ export default class AutoUpdates extends React.Component {
+                                   onClick={ this.handleChange }>
+                               {_("Save changes")}
+                           </Button>
+                           <Button variant="link" className="btn-cancel" onClick={() => this.setState({ showModal: false })}>
```
btn-cancel class not used anywhere. You can just drop it

skobyda replied:
Fixed

garrett commented:
Rebased and added fixes for:

1. Layout cards (using gallery, not grid, as PF "grid" is intended to be a 12-column Bootstrap replacement)
2. Smaller auto apply updates dialog

garrett commented:
I don't see content for the update history, BTW. Of course, I've never updated in this VM, but still... either there needs to be an empty state, or we don't show update history when there is no update history (which is what I'd prefer).

skobyda commented:
Thanks @garrett for fixes. So I added a condition so the history card is not rendered at all when there is no history.
@KKoukiou and @andreasn, your issues should be fixed, so it's ready for re-review.

andreasn commented:
Generally looks good, but it seems it's getting this layout if wider than 1616px
![image](https://user-images.githubusercontent.com/209895/100218442-0e8ef280-2f15-11eb-8483-ce8516c00aee.png)

andreasn commented:
Maybe only **Status** and **Automatic updates** should be using gallery and then **Available updates** would go outside that?

andreasn commented:
Katerina suggested taking a look at how the virtual machines page does it's layout. All of the cards are in a section, but the top two cards are in their own div, and the listing is in an article tag.

garrett commented:
I changed the layout to use flex (CSS); perhaps it can use PF's flex for the main part (although it'd still need to have flex-basis set for the individual elements). I basically set that as the target width (via flex-basis) and stretch them to meet the content area otherwise.

andreasn reviewed (approved):
UI and HTML/CSS looks good to me!

KKoukiou commented on pkg/packagekit/updates.jsx:
```diff
@@ -509,6 +514,116 @@ const AskRestart = ({ onIgnore, onRestart, history }) => <>
+                title: _("Status"),
+                className: "pk-card-info",
+                actions: (<Tooltip content={_("Check for updates")}>
+                    <Button variant="secondary" onClick={this.props.handleRefresh}><i className="fa fa-refresh" /></Button>
```
Can you please use PF4 icons here?
It will make the migration process easier.

skobyda replied:
I could not find a same icon in PF4, But I found spinner2 (https://pf4.patternfly.org/icons/) `<i class="pf-icon pf-icon-spinner2"></i>`, Maybe that icon is okay too

skobyda replied:
So I used RedoIcon in the end. Fixed

KKoukiou commented on pkg/packagekit/autoupdates.jsx:
```diff
@@ -320,41 +380,54 @@ export default class AutoUpdates extends React.Component {
+        );
+
+        return (<>
+            <Button variant="primary"
```
PF4 documentation suggests that there should only be one primary button per page, I would make this secondary.

skobyda replied:
Fixed

KKoukiou reviewed (changes requested):
Some minor comments and a design suggestion. 
I really don't like how the first two cards take all the vertical space, and the table needs to squeeze there on the right. 
I would prefer to see something similar of what we have in the machines main page. Two short cards above, and the one card with the table bellow. We can even reuse the CSS from there.

KKoukiou commented:
So now I see:
![Screen Shot 2020-12-01 at 11 18 56](https://user-images.githubusercontent.com/14921356/100728329-b4d07180-33c7-11eb-93d2-32144b3a1b02.png)


I would really prefer to see:
![Screen Shot 2020-12-01 at 11 23 11](https://user-images.githubusercontent.com/14921356/100728348-bb5ee900-33c7-11eb-9ea5-cee23e56dbec.png)


And ideally even without the software updates header on top, it's pretty useless.

skobyda commented on pkg/packagekit/updates.jsx:
```diff
@@ -509,6 +514,116 @@ const AskRestart = ({ onIgnore, onRestart, history }) => <>
+
+    const icon = highestSeverity ? <span className={PK.getSeverityIcon(highestSeverity)} /> : <CheckIcon color="green" />;
+
+    return (<Flex>
```
Yes. To show it a bit more abstractly, I need to make a Flex of 3 items (A,B,C) in a hierarchy.
```
      X
A         Y
        B     C
```
Where on first level, I want a Flex for a A and Y, and for the group Y made out of B and C I want a different Flex. But maybe that's not a correct solution to do that, I'm open to new ideas.

skobyda commented on pkg/packagekit/autoupdates.jsx:
```diff
@@ -320,41 +380,53 @@ export default class AutoUpdates extends React.Component {
-            </div>);
+                    </FormGroup>
+
+                    <Alert variant="info" title={_("This host will reboot after updates.")} isInline />
```
You are right, it kinda misses the context and one could think that the host will update after you change the configuration of autoupdates. I changed it to "This host will reboot after updates are installed".

skobyda commented on pkg/packagekit/autoupdates.jsx:
```diff
@@ -320,41 +380,54 @@ export default class AutoUpdates extends React.Component {
+                    }}>
+                {_("Edit")}
+            </Button>
+            <Modal position="top" variant="small" id="automatic-updates-dialog" isOpen={this.state.showModal} onClose={() => this.setState({ showModal: false })}
```
Fixed

skobyda commented:
> For the screenshot I just edit the pk-overview grid to: --pf-l-gallery--GridTemplateColumns: repeat(auto-fit,minmax(0.5fr,1fr));

Fixed

garrett reviewed (changes requested):
Oh, this PR is back to breaking the layout again. This time, it's mobile sizes.

I had a fix that covered mobile and desktop sizes, but I guess it was reverted? (Or did I forget to push it up?)

skobyda commented:
> Oh, this PR is back to breaking the layout again. This time, it's mobile sizes.
> 
> I had a fix that covered mobile and desktop sizes, but I guess it was reverted? (Or did I forget to push it up?)

Or maybe I accidentally force pushed over them, sorry :/ . Can you push them again? Or give me a link for your branch?

garrett commented:
Meanwhile, I realized my nice and simple content-based layout fix was broken in WebKit, so I'm redoing it to work cross-browser using grid and breakpoints instead of working on content sizing.

garrett commented:
I've pushed my revised changes. I might've gotten the breakpoint wrong, but I think it's probably right. I'm rebuilding a VM now to confirm, instead of scp-ing local files to a VM (which is buggggggy)... but it takes half a hour each time I rebuild a VM to test (and then fix the DNS inside... sigh).

Anyway, it's late here and I wanted to share the fix, even if the breakpoint might not be 100% correct.

This should hopefully handle desktop sizes and mobile sizes just fine and work cross-browser, unlike all the other solutions (which had one problem or another).

garrett commented:
Is the expanded item changelog missing the markup-ifier? Or is the CSS broken for it? Things look messier than I remember them looking.

garrett reviewed (approved):
LGTM now.

I see a few minor things, but nothing that needs to be in this PR. It's tiny stuff for follow-ups, like:
- proper spacing between the icon and text
- fixing a regression with the changelog text that apparently happened a while ago
- and a few other minor things that are probably debatable

But none of this is the result of the PR whatsoever.

What's in the PR looks great to me! Looking forward to this landing soon.

KKoukiou reviewed (commented):
Please add an up to date screenshot for the release notes and fix the tests.

KKoukiou commented:
One test failure which seems related. Retrying for comparison: 
https://logs.cockpit-project.org/logs/pull-14934-20201207-011005-2adbe315-rhel-8-3-distropkg/log.html#296

KKoukiou commented on pkg/packagekit/updates.scss:
```diff
@@ -351,3 +357,19 @@ table.header-buttons {
 table.available {
```
This needs to be deleted.
 The margin right now is too much:
![Screen Shot 2020-12-07 at 10 12 04](https://user-images.githubusercontent.com/14921356/101331781-bccf5c00-3874-11eb-875c-498e85c6551d.png)

KKoukiou replied:
This needs to be deleted.
 The margin right now is too much:
![Screen Shot 2020-12-07 at 10 12 04](https://user-images.githubusercontent.com/14921356/101331781-bccf5c00-3874-11eb-875c-498e85c6551d.png)

KKoukiou commented on pkg/packagekit/updates.jsx:
```diff
@@ -509,6 +515,116 @@ const AskRestart = ({ onIgnore, onRestart, history }) => <>
+                        <CardTitle><h2>{card.title}</h2></CardTitle>
+                        {card.actions && <CardActions>{card.actions}</CardActions>}
+                    </CardHeader>
+                    <CardBody>
```
This needs the class 'contains-list'
In all other pages when we use tables inside card bodes we remove the margins.
I don't have an opinion if it's better or worse design, but it's really important that we are consistent in the usage of the cards.
So add here className='contains-list' and the result should be like the following screenshot (after your also remove the CSS rule with the added margin)

![Screen Shot 2020-12-07 at 10 14 23](https://user-images.githubusercontent.com/14921356/101332180-3ebf8500-3875-11eb-9c2e-ca9d0d0f28f9.png)

KKoukiou reviewed (changes requested):
Some last design changes, the code looks okay!

Two other comments inline and one more here.
So I know you did not add it, but this needs to be removed,  in the updates-log CSS class remove padding-top: 8em;
    padding-top: 8em;

With this padding:

![Screen Shot 2020-12-07 at 10 20 30](https://user-images.githubusercontent.com/14921356/101332859-2bf98000-3876-11eb-8eac-19eee50a878c.png)

Without the padding:
![Screen Shot 2020-12-07 at 10 23 40](https://user-images.githubusercontent.com/14921356/101332965-4cc1d580-3876-11eb-959a-a77f29f64366.png)

I really believe this padding takes up space without offering any UI improvement.

KKoukiou commented:
@skobyda also one tests is failing consistently: https://logs.cockpit-project.org/logs/pull-14934-20201207-083443-2adbe315-rhel-8-3-distropkg/log.html

KKoukiou commented:
@skobyda since you will repost anyway, please also remove the extra padding added in flow-list, it makes the content of the table too much padded.
Before:

![Screen Shot 2020-12-07 at 10 46 56](https://user-images.githubusercontent.com/14921356/101335540-a1b31b00-3879-11eb-8226-a6749b46dc8d.png)

After removing the extra padding:
https://github.com/cockpit-project/cockpit/blob/2adbe3154995882695bd76b33f59d6be8ab5c40d/pkg/packagekit/updates.scss#L260
![Screen Shot 2020-12-07 at 10 46 38](https://user-images.githubusercontent.com/14921356/101335545-a1b31b00-3879-11eb-9e22-f3612732bd47.png)

skobyda commented:
So I fixed the css issues, and fixed the one failing test (hopefully I didn't introduce new failure)
Here's how it looks now:
![Screenshot from 2020-12-07 12-44-54](https://user-images.githubusercontent.com/42733240/101348102-691c3d00-388b-11eb-851b-7bbadcb0e825.png)
![Screenshot from 2020-12-07 12-42-28](https://user-images.githubusercontent.com/42733240/101348104-69b4d380-388b-11eb-9f4d-994d1b27ee65.png)
![Screenshot from 2020-12-07 12-49-20](https://user-images.githubusercontent.com/42733240/101348107-69b4d380-388b-11eb-9c14-7fef84ee5876.png)

garrett commented:
I force-pushed a quick fix to "contains-list'", to make it "contains-list" (without the `'`).

garrett commented:
Oof, that broke the other cards, as apparently they're all using the same class. Whoops.

KKoukiou commented on pkg/packagekit/updates.jsx:
```diff
@@ -509,6 +514,117 @@ const AskRestart = ({ onIgnore, onRestart, history }) => <>
+            cardContents.push({
+                id: "update-history",
+                title: _("Update history"),
+                body: <History packagekit={this.props.history} />
```
This needs also contains-list I think

skobyda replied:
Fixed

garrett commented:
While here, I made a fix for the icon size and alignment. There should _never_ be text inside of a pficon / font-awesome icon element. And the alignment / spacing is a bit off, so I fixed it too.

garrett commented:
BTW: I think we usually use "packagekit: " as the prefix, as it's the name of the package: "cockpit-packagekit". We should probably make sure the commit(s) are named appropriately.

garrett commented:
(The icons would probably be better ported to PF4 icons... except we're probably wind up with a whole lot of repeated inline SVGs.)

marusak commented on pkg/packagekit/autoupdates.jsx:
```diff
@@ -249,67 +248,128 @@ function getBackend(forceReinit) {
+    let str;
+    if (enabled) {
+        str = type == "security" ? _("Security updates ") : _("Updates ");
+        str += day == "" ? cockpit.format(_("will be applied $0 at $1"), days[day], time) : cockpit.format(_("will be applied $0 at $1"), days[day], time);
```
Whats the difference between the first and the second string?

skobyda replied:
Oh, they were different at first. Then I ammended them without realizing both strings are the same now. Will fix it

skobyda commented:
> BTW: I think we usually use "packagekit: " as the prefix, as it's the name of the package: "cockpit-packagekit". We should probably make sure the commit(s) are named appropriately.

fixed

KKoukiou reviewed (approved):
Thanks, 
this is fine as is.
On a followup pls remove the spinner from he cancel button.
![Screen Shot 2020-12-08 at 14 53 03](https://user-images.githubusercontent.com/14921356/101492863-8d4c4c80-3965-11eb-9ec6-1634c06e95d7.png)
